File: guest/hello-ts/assembly/impl.ts
// impl.ts — the application half of this guest: define the exported
// functions below (assembly/gen/bindings.ts imports them by name).
// crabgen scaffolds this file ONCE and never overwrites it; `crabgen
// regen` prints any missing function signatures instead of editing it.
import {
  GreetRequest,
  ResString,
  ResU32,
} from "./gen/bindings";

// greet handles crab:hello-ts/greeter@0.1.0#greet.
// A non-null .err is a function-level failure (status-1 reply).
export function greet(req: GreetRequest): ResString {
  const bang = req.excited !== null && req.excited!.value ? "!!!" : "!";
  return ResString.ok("Hello from TS, " + req.name + bang);
}

// add handles crab:hello-ts/greeter@0.1.0#add.
// add(a: u32, b: u32) -> u32 (u32 arithmetic wraps naturally)
export function add(a: u32, b: u32): ResU32 {
  return ResU32.ok(a + b);
}

File: guest/hello-cpp/impl.cpp
// impl.cpp — the application half of this guest: define the impl:: functions
// declared in gen/bindings.hpp. crabgen scaffolds this file ONCE and never
// overwrites it; `crabgen regen` prints any missing function signatures
// instead of editing it (a missing definition is also a LINK ERROR naming
// the symbol when build.sh links the module).
#include "gen/bindings.hpp"

namespace impl {

// greet handles crab:hello-cpp/greeter@0.1.0#greet.
// A non-empty .err is a function-level failure (status-1 reply).
crab::Res<std::string> greet(gen::GreetRequest req) {
  const char* bang = (req.excited.has_value() && *req.excited) ? "!!!" : "!";
  return crab::Res<std::string>{"Hello from C++, " + req.name + bang, {}};
}

// add handles crab:hello-cpp/greeter@0.1.0#add.
// add(a: u32, b: u32) -> u32 (unsigned arithmetic wraps naturally)
crab::Res<uint32_t> add(uint32_t a, uint32_t b) {
  return crab::Res<uint32_t>{a + b, {}};
}

}  // namespace impl
